source/val/validate_graph.cpp

```diff
@@ -0,0 +1,605 @@
+
+    uint32_t interface_ptr_type = interface_inst->type_id();
+    auto interface_ptr_inst = _.FindDef(interface_ptr_type);
+    auto interface_pointee_type = interface_ptr_inst->word(3);
+    if (interface_pointee_type != corresponding_graph_io_type) {
+      return _.diag(SPV_ERROR_INVALID_DATA, inst)
+             << spvOpcodeString(inst->opcode()) << " Interface ID type "
+             << _.getIdName(interface_pointee_type)
+             << " must match the type of the corresponding graph I/O "
+             << _.getIdName(corresponding_graph_io_type);
+    }
+  }
+
+  return SPV_SUCCESS;
+}
+
+spv_result_t ValidateGraph(ValidationState_t& _, const Instruction* inst) {
+  // Result Type must be an OpTypeGraphARM
+  if (!IsGraphType(_, inst->type_id())) {
+    return _.diag(SPV_ERROR_INVALID_DATA, inst)
+           << spvOpcodeString(inst->opcode())
+           << " Result Type must be an OpTypeGraphARM.";
+  }
+
+  return SPV_SUCCESS;
+}
+
+spv_result_t ValidateGraphInput(ValidationState_t& _, const Instruction* inst) {
+  // Check type of InputIndex
+  auto input_index_inst = _.FindDef(inst->GetOperandAs<uint32_t>(2));
+  if (!input_index_inst ||
+      !_.IsIntScalarType(input_index_inst->type_id(), 32)) {
+    return _.diag(SPV_ERROR_INVALID_DATA, inst)
+           << spvOpcodeString(inst->opcode())
+           << " InputIndex must be a 32-bit integer.";
+  }
+
+  bool has_element_index = inst->words().size() > 4;
+
+  // Check type of ElementIndex
+  if (has_element_index) {
+    auto element_index_inst = _.FindDef(inst->GetOperandAs<uint32_t>(3));
+    if (!element_index_inst ||
+        !_.IsIntScalarType(element_index_inst->type_id(), 32)) {
+      return _.diag(SPV_ERROR_INVALID_DATA, inst)
+             << spvOpcodeString(inst->opcode())
+             << " ElementIndex must be a 32-bit integer.";
+    }
+  }
+
+  // Check the instruction is not the first in the module
+  size_t inst_num = inst->LineNum() - 1;
+  if (inst_num == 0) {
+    return _.diag(SPV_ERROR_INVALID_LAYOUT, inst)
+           << spvOpcodeString(inst->opcode())
+           << " cannot be the first instruction.";
+  }
+
+  // Check the instruction is preceded by OpGraphARM or OpGraphInputARM
+  auto previous_inst = &_.ordered_instructions()[inst_num - 1];
+  if ((previous_inst->opcode() != spv::Op::OpGraphARM) &&
+      (previous_inst->opcode() != spv::Op::OpGraphInputARM)) {
+    return _.diag(SPV_ERROR_INVALID_DATA, inst)
+           << "Op" << spvOpcodeString(inst->opcode())
+           << " must immediately follow an OpGraphARM or OpGraphInputARM "
+              "instruction.";
+  }
+
+  // Find graph definition
+  auto graph_inst = &_.ordered_instructions()[inst_num];
+  while (--inst_num) {
+    graph_inst = &_.ordered_instructions()[inst_num];
+    if (graph_inst->opcode() == spv::Op::OpGraphARM) {
+      break;
+    }
+  }
+
+  // Can the InputIndex be evaluated?
+  // If not, there's nothing more we can validate here.
```

Comment:
Validate that it is a 32-bit integer as in the spec? Same goes for element index.

Reply:
These checks are already present at the beginning of the function and there are corresponding unit tests.